test/common/event/guarddog_impl_test.cc

```diff
@@ -0,0 +1,152 @@
+
+  void startNeglectedDogAndWait() {
+    InSequence s;
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    GuardDogImpl gd(fakestats_, config_kill_, time_source_);
+    auto unpet_dog = gd.getWatchDog(0);
+    gd.force_check();
+    time_point_ += std::chrono::milliseconds(500);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    gd.force_check();
+  }
+
+  void startNeglectedDogsAndWait() {
+    InSequence s;
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    GuardDogImpl gd(fakestats_, config_multikill_, time_source_);
+    auto unpet_dog1 = gd.getWatchDog(0);
+    gd.force_check();
+    auto unpet_dog2 = gd.getWatchDog(1);
+    gd.force_check();
+    time_point_ += std::chrono::milliseconds(501);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    gd.force_check();
+  }
+
+  void startNeglectedDogAndTouchedDog() {
+    InSequence s;
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    GuardDogImpl gd(fakestats_, config_multikill_, time_source_);
+    auto unpet_dog = gd.getWatchDog(0);
+    auto pet_dog = gd.getWatchDog(1);
+    for (int i = 0; i < 6; i++) {
+      time_point_ += std::chrono::milliseconds(100);
+      EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+      pet_dog->touch();
+      gd.force_check();
+    }
+  }
+
+  testing::NiceMock<Server::Configuration::MockMain> config_kill_;
+  testing::NiceMock<Server::Configuration::MockMain> config_multikill_;
+  testing::NiceMock<Stats::MockStore> fakestats_;
+  MockSystemTimeSource time_source_;
+  std::chrono::system_clock::time_point time_point_;
+};
+
+class GuardDogMissTest : public testing::Test {
+protected:
+  GuardDogMissTest() : config_miss(1, 1000, 0, 0), config_mega(1000, 1, 0, 0) {}
+
+  void startNeglectedDogAndWait(bool mega) {
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    GuardDogImpl gd(stats_store, mega ? config_mega : config_miss, time_source_);
+    auto unpet_dog = gd.getWatchDog(0);
+    time_point_ += std::chrono::milliseconds(501);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    gd.force_check();
+    gd.stopWatching(unpet_dog);
+    unpet_dog = nullptr;
+  }
+
+  testing::NiceMock<Server::Configuration::MockMain> config_miss;
+  testing::NiceMock<Server::Configuration::MockMain> config_mega;
+  testing::NiceMock<Stats::MockStore> stats_store;
+  MockSystemTimeSource time_source_;
+  std::chrono::system_clock::time_point time_point_;
+};
+
+TEST_F(GuardDogDeathTest, KillDeathTest) { EXPECT_DEATH(startNeglectedDogAndWait(), ""); }
+TEST_F(GuardDogDeathTest, MultiKillDeathTest) { EXPECT_DEATH(startNeglectedDogsAndWait(), ""); }
```

Comment:
See comment below about inlining stuff into the `TEST_F`. Also, those method names are _very_ similar, I had to stare for a while to figure out these two tests were different (but, my glasses are currently at the optometrist undergoing a change in lens, so that might be it as well).